### PR TITLE
sql: Add hint to DROP with no CASCADE errors

### LINKED ISSUE #21942: Add a hint to use CASCADE for dropping objects with dependencies
When a user tries to drop an object with dependencies, we return an error 
```
staging > drop cluster auction_house;
ERROR:  cannot drop cluster with active objects

staging > drop source demo;
ERROR:  cannot drop source demo: still depended upon by view on_time_bids
```
Add a hint suggesting to use `CASCADE` to drop the object and all downstream dependencies.
This is postgres's hint: `Use DROP ... CASCADE to drop the dependent objects too.`

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds a hint to all errors that result from failed DROPs due to dependent objects still existing and not using CASCADE. The hint says to use CASCADE to drop dependent objects.

Resolves #21942

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/sql/src/plan/error.rs [hunk 1]
```diff
@@ -474,6 +480,13 @@ impl fmt::Display for PlanError {
             Self::DropLastSubsource { source } => write!(f, "SOURCE {} must retain at least one non-progress subsource", source.quoted()),
             Self::DropProgressCollection { progress_collection, source: _} => write!(f, "SOURCE {} is a progress collection and cannot be dropped independently of its primary source", progress_collection.quoted()),
             Self::DropNonSubsource { non_subsource, source} => write!(f, "SOURCE {} is a not a subsource of {}", non_subsource.quoted(), source.quoted()),
+            Self::DependentObjectsStillExist {object_type, object_name, dependents} => {
+                let reason = match &dependents[..] {
+                    [] => " because other objects depend on it".to_string(),
+                    dependents => format!(": still depended upon by {}", dependents.join(", ")),
+                };
+                write!(f, "cannot drop {object_type} {object_name}{reason}")
```
Comment (ParkMyCar) [inline src/sql/src/plan/error.rs:488]: `object_name` and all entries in `dependents` should be quoted

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -4136,10 +4136,15 @@ fn plan_drop_role(
...
-                        );
+                        return Err(PlanError::DependentObjectsStillExist {
+                            object_type: "role".to_string(),
+                            object_name: name.as_str().to_string(),
+                            dependents: vec![format!(
+                                "membership of role {} in role {}",
+                                role.name(),
+                                member_role.name()
+                            )],
+                        });
```
Comment (ParkMyCar) [inline src/sql/src/plan/statement/ddl.rs:4147]: can you add a test case for this?

File: test/sqllogictest/object_ownership.slt [hunk 3]
```diff
@@ -2713,7 +2713,8 @@ COMPLETE 0
 simple conn=mz_system,user=mz_system
 DROP OWNED BY materialize;
 ----
-db error: ERROR: cannot drop "materialize.public.t" without CASCADE: still depended upon by non-owned catalog items "materialize.public.v"
+db error: ERROR: cannot drop table "materialize.public.t": still depended upon by "materialize.public.v"
```
Comment (ParkMyCar) [inline test/sqllogictest/object_ownership.slt:2716]: Can you add a test case where it's depended on by multiple objects?

**Review by ParkMyCar — APPROVED** [review]
Nice!!

File: src/sql/src/plan/statement/ddl.rs (hunk 2 above)
Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs]: Actually, I just realized that there is no `CASCADE` for `DROP ROLE` so I'm reverting this line.

File: src/sql/src/plan/error.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql/src/plan/error.rs]: Done

File: test/sqllogictest/object_ownership.slt (hunk 3 above)
Comment (jkosh44) [reply test/sqllogictest/object_ownership.slt]: Done

**Pull request merged**